File: examples/batch-processor.py
#!/usr/bin/env python3
"""
Drupal Content Batch Processor (Python)

Example client-side script demonstrating batch processing and aggregation
of Drupal content across multiple pages.

Usage: python3 examples/batch-processor.py

Dependencies: pip install requests
"""

import requests
import sys
from typing import List, Dict, Any, Optional, Callable


class DrupalContentBatchProcessor:
    def __init__(self, base_url: str = 'http://localhost:3000'):
        self.base_url = base_url.rstrip('/')

    def fetch_content_page(self, page: int = 1, limit: int = 50, content_type: Optional[str] = None) -> Dict[str, Any]:
        """Fetch content with pagination support"""
        params = {
            'page': str(page),
            'limit': str(limit)
        }

        if content_type:
            params['type'] = content_type

        url = f"{self.base_url}/content"
        print(f"Fetching: {url} with params {params}")

        response = requests.get(url, params=params)
        response.raise_for_status()

        data = response.json()
        content_count = len(data.get('content', [])) if data.get('content') else 0
        print(f"Response for page {page}: success={data.get('success')}, content items={content_count}")
        return data

    def get_pagination_info(self, content_type: Optional[str] = None) -> Dict[str, Any]:
        """Get pagination information from the first page"""
        first_page = self.fetch_content_page(1, 1, content_type)
        pagination = first_page.get('pagination', {})

        return {
            'totalPages': pagination.get('totalPages', 0),
            'totalItems': pagination.get('totalItems', 0),
            'hasNextPage': pagination.get('hasNextPage', False)
        }

    def fetch_items(self, count: int, options: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Fetch a specific number of items, aggregating across multiple pages"""
        if options is None:
            options = {}

        content_type = options.get('type')
        page_size = options.get('pageSize', 50)
        on_progress = options.get('onProgress')

        print(f"Fetching {count} items (page size: {page_size}, type: {content_type or 'all'})")

        # Get pagination info
        pagination = self.get_pagination_info(content_type)
        print(f"Site has {pagination['totalPages']} pages total")

        # Calculate how many pages we need
        pages_needed = (count + page_size - 1) // page_size  # Ceiling division
        actual_pages = min(pages_needed, pagination['totalPages'])

        print(f"Need to fetch {actual_pages} pages to get {count} items")

        # Fetch pages sequentially to avoid overwhelming the server
        results = []
        for page in range(1, actual_pages + 1):
            try:
                data = self.fetch_content_page(page, page_size, content_type)
                if not data.get('success'):
                    error_msg = data.get('error', 'Unknown error')
                    raise Exception(f"Failed to fetch page {page}: {error_msg}")

                if on_progress:
                    content_length = len(data.get('content', []))
                    on_progress(page, actual_pages, content_length)

                results.append(data)

            except Exception as error:
                print(f"Error fetching page {page}: {str(error)}", file=sys.stderr)
                raise

        # Aggregate all content
        all_content = []
        for result in results:
            if result and result.get('content') and isinstance(result['content'], list):
                all_content.extend(result['content'])
            else:
                print(f"Warning: Skipping result with invalid content: {result}", file=sys.stderr)

        # Trim to exact count requested
        final_content = all_content[:count]

        return {
            'content': final_content,
            'totalFetched': len(all_content),
            'requested': count,
            'actual': len(final_content),
            'pagesFetched': len(results),
            'pagination': pagination
        }

    def example_fetch_75_items(self) -> Dict[str, Any]:
        """Example: Fetch 75 most recent items (demonstrates cross-page aggregation)"""
        print("=== Example: Fetch 75 Most Recent Items ===")
        print("Site paginates by 50 items per page, so we need 2 pages\n")

        def progress_callback(page: int, total_pages: int, items_in_page: int):
            print(f"Fetched page {page}/{total_pages} ({items_in_page} items)")

        try:
            result = self.fetch_items(75, {
                'pageSize': 50,
                'onProgress': progress_callback
            })

            print("\n=== Results ===")
            print(f"Requested: {result['requested']} items")
            print(f"Fetched: {result['totalFetched']} items from {result['pagesFetched']} pages")
            print(f"Returned: {result['actual']} items (trimmed to request)")

            # Show sample of results
            print("\n=== Sample Content ===")
            for i, item in enumerate(result['content'][:3], 1):
                print(f"{i}. {item.get('title', 'No title')} ({item.get('type', 'Unknown')}) - {item.get('author', 'Unknown')}")

            if len(result['content']) > 3:
                print(f"... and {len(result['content']) - 3} more items")

            return result

        except Exception as error:
            print(f"Error fetching items: {str(error)}", file=sys.stderr)
            raise

    def example_fetch_75_items_demonstration(self) -> Dict[str, Any]:
        """Example: Fetch 75 most recent items (demonstrates cross-page aggregation)
        NOTE: This is a demonstration that shows the logic. In a real implementation,
        you would handle browser session management and rate limiting.
        """
        print("=== Example: Fetch 75 Most Recent Items (Demonstration) ===")
        print("Site paginates by 50 items per page, so we need 2 pages\n")

        # Simulate getting pagination info
        print("Step 1: Get pagination information from first page")
        print("API Call: GET /content?page=1&limit=1")
        print("Response: 46 total pages detected\n")

        total_pages = 46  # From our API testing
        items_needed = 75
        page_size = 50
        pages_needed = (items_needed + page_size - 1) // page_size  # Ceiling division

        print("Step 2: Calculate requirements")
        print(f"- Need {items_needed} items total")
        print(f"- Site pages by {page_size} items per page")
        print(f"- Need to fetch {pages_needed} pages ({pages_needed * page_size} items)")
        print(f"- Will trim results to exactly {items_needed} items\n")

        print("Step 3: Fetch pages sequentially")
        total_fetched = 0
        all_content = []

        for page in range(1, pages_needed + 1):
            print(f"API Call: GET /content?page={page}&limit={page_size}")
            print(f"Response: {page_size} items from page {page}")
            total_fetched += page_size

            # Simulate content items
            for i in range(1, page_size + 1):
                item_number = ((page - 1) * page_size) + i
                all_content.append({
                    'id': 10000 + item_number,
                    'title': f'Content Item {item_number}',
                    'type': 'Event' if page == 1 else 'Article',
                    'author': f'Author {item_number % 5 + 1}',
                    'status': 'Published'
                })

        print("\nStep 4: Aggregate and trim results")
        print(f"- Fetched {total_fetched} items from {pages_needed} pages")
        print(f"- Trimming to {items_needed} items as requested")

        final_content = all_content[:items_needed]

        print("\n=== Results ===")
        print(f"Requested: {items_needed} items")
        print(f"Fetched: {total_fetched} items from {pages_needed} pages")
        print(f"Returned: {len(final_content)} items (trimmed to request)")

        # Show sample of results
        print("\n=== Sample Content ===")
        for i, item in enumerate(final_content[:5], 1):
            print(f"{i}. {item['title']} ({item['type']}) - {item['author']}")

        if len(final_content) > 5:
            print(f"... and {len(final_content) - 5} more items")

        return {
            'content': final_content,
            'totalFetched': total_fetched,
            'requested': items_needed,
            'actual': len(final_content),
            'pagesFetched': pages_needed
        }

    def example_fetch_all_of_type(self, content_type: str) -> Dict[str, Any]:
        """Example: Fetch all items of a specific type"""
        print(f"=== Example: Fetch All {content_type} Items ===")

        try:
            # First get pagination info for this type
            pagination = self.get_pagination_info(content_type)
            print(f"Found {pagination['totalPages']} pages of {content_type} content")

            result = self.fetch_items(
                pagination['totalPages'] * 50,  # Estimate based on pages
                {'type': content_type}
            )

            print(f"Fetched {result['actual']} {content_type} items total")
            return result

        except Exception as error:
            print(f"Error fetching {content_type} items: {str(error)}", file=sys.stderr)
            raise


def main():
    """CLI interface"""
    processor = DrupalContentBatchProcessor()

    try:
        # Example 1: Fetch 75 items (crosses page boundary) - Demonstration
        processor.example_fetch_75_items_demonstration()

        print("\n" + "=" * 50 + "\n")

        # Example 2: Show how it would work with real API
        print("=== Real Implementation Notes ===")
        print("To run with actual API calls:")
        print("1. Ensure Drupal session is loaded: POST /login/load")
        print("2. Run: processor.example_fetch_75_items()")
        print("3. Handle potential browser session rate limiting")
        print("4. Consider adding delays between requests")

    except Exception as error:
        print(f"Batch processing demonstration failed: {str(error)}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: examples/batch-processor.js
#!/usr/bin/env node

/**
 * Drupal Content Batch Processor
 *
 * Example client-side script demonstrating batch processing and aggregation
 * of Drupal content across multiple pages.
 *
 * Usage: node examples/batch-processor.js
 * Requires: Node.js 18+ (uses built-in fetch)
 */

class DrupalContentBatchProcessor {
  constructor(baseUrl = 'http://localhost:3000') {
    this.baseUrl = baseUrl;
  }

  /**
   * Fetch content with pagination support
   */
  async fetchContentPage(page = 1, limit = 50, type = null) {
    const params = new URLSearchParams({
      page: page.toString(),
      limit: limit.toString()
    });

    if (type) {
      params.append('type', type);
    }

    const url = `${this.baseUrl}/content?${params}`;
    console.log(`Fetching: ${url}`);

    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    const data = await response.json();
    console.log(`Response for page ${page}: success=${data.success}, content items=${data.content ? data.content.length : 'undefined'}`);
    return data;
  }

  /**
   * Get pagination information from the first page
   */
  async getPaginationInfo(type = null) {
    const firstPage = await this.fetchContentPage(1, 1, type);
    return {
      totalPages: firstPage.pagination.totalPages,
      totalItems: firstPage.pagination.totalItems,
      hasNextPage: firstPage.pagination.hasNextPage
    };
  }

  /**
   * Fetch a specific number of items, aggregating across multiple pages
   */
  async fetchItems(count, options = {}) {
    const {
      type = null,
      pageSize = 50,
      onProgress = null
    } = options;

    console.log(`Fetching ${count} items (page size: ${pageSize}, type: ${type || 'all'})`);

    // Get pagination info
    const pagination = await this.getPaginationInfo(type);
    console.log(`Site has ${pagination.totalPages} pages total`);

    // Calculate how many pages we need
    const pagesNeeded = Math.ceil(count / pageSize);
    const actualPages = Math.min(pagesNeeded, pagination.totalPages);

    console.log(`Need to fetch ${actualPages} pages to get ${count} items`);

    // Create page fetch promises
    const pagePromises = [];
    for (let page = 1; page <= actualPages; page++) {
      const promise = this.fetchContentPage(page, pageSize, type)
        .then(data => {
          if (!data.success) {
            throw new Error(`Failed to fetch page ${page}: ${data.error || 'Unknown error'}`);
          }
          if (onProgress) {
            const contentLength = data.content ? data.content.length : 0;
            onProgress(page, actualPages, contentLength);
          }
          return data;
        })
        .catch(error => {
          console.error(`Error fetching page ${page}:`, error.message);
          throw error;
        });
      pagePromises.push(promise);
    }

    // Fetch pages with controlled concurrency
    const results = [];
    for (let i = 0; i < pagePromises.length; i += 1) {
      const batch = pagePromises.slice(i, i + 1);
      const batchResults = await Promise.all(batch);
      results.push(...batchResults);
    }

    // Aggregate all content
    const allContent = [];
    for (const result of results) {
      if (result && result.content && Array.isArray(result.content)) {
        allContent.push(...result.content);
      } else {
        console.warn('Skipping result with invalid content:', result);
      }
    }

    // Trim to exact count requested
    const finalContent = allContent.slice(0, count);

    return {
      content: finalContent,
      totalFetched: allContent.length,
      requested: count,
      actual: finalContent.length,
      pagesFetched: results.length,
      pagination: pagination
    };
  }

  /**
   * Example: Fetch 75 most recent items (demonstrates cross-page aggregation)
   */
  async exampleFetch75Items() {
    console.log('=== Example: Fetch 75 Most Recent Items ===');
    console.log('Site paginates by 50 items per page, so we need 2 pages\n');

    const progressCallback = (page, totalPages, itemsInPage) => {
      console.log(`Fetched page ${page}/${totalPages} (${itemsInPage} items)`);
    };

    try {
      const result = await this.fetchItems(75, {
        pageSize: 50,
        onProgress: progressCallback
      });

      console.log('\n=== Results ===');
      console.log(`Requested: ${result.requested} items`);
      console.log(`Fetched: ${result.totalFetched} items from ${result.pagesFetched} pages`);
      console.log(`Returned: ${result.actual} items (trimmed to request)`);

      // Show sample of results
      console.log('\n=== Sample Content ===');
      result.content.slice(0, 3).forEach((item, index) => {
        console.log(`${index + 1}. ${item.title} (${item.type}) - ${item.author}`);
      });

      if (result.content.length > 3) {
        console.log(`... and ${result.content.length - 3} more items`);
      }

      return result;

    } catch (error) {
      console.error('Error fetching items:', error.message);
      throw error;
    }
  }

  /**
   * Example: Fetch 75 most recent items (demonstrates cross-page aggregation)
   * NOTE: This is a demonstration that shows the logic. In a real implementation,
   * you would handle browser session management and rate limiting.
   */
  async exampleFetch75ItemsDemonstration() {
    console.log('=== Example: Fetch 75 Most Recent Items (Demonstration) ===');
    console.log('Site paginates by 50 items per page, so we need 2 pages\n');

    // Simulate getting pagination info
    console.log('Step 1: Get pagination information from first page');
    console.log('API Call: GET /content?page=1&limit=1');
    console.log('Response: 46 total pages detected\n');

    const totalPages = 46; // From our API testing
    const itemsNeeded = 75;
    const pageSize = 50;
    const pagesNeeded = Math.ceil(itemsNeeded / pageSize);

    console.log(`Step 2: Calculate requirements`);
    console.log(`- Need ${itemsNeeded} items total`);
    console.log(`- Site pages by ${pageSize} items per page`);
    console.log(`- Need to fetch ${pagesNeeded} pages (${pagesNeeded * pageSize} items)`);
    console.log(`- Will trim results to exactly ${itemsNeeded} items\n`);

    console.log('Step 3: Fetch pages sequentially');
    let totalFetched = 0;
    const allContent = [];

    for (let page = 1; page <= pagesNeeded; page++) {
      console.log(`API Call: GET /content?page=${page}&limit=${pageSize}`);
      console.log(`Response: ${pageSize} items from page ${page}`);
      totalFetched += pageSize;

      // Simulate content items
      for (let i = 1; i <= pageSize; i++) {
        const itemNumber = ((page - 1) * pageSize) + i;
        allContent.push({
          id: 10000 + itemNumber,
          title: `Content Item ${itemNumber}`,
          type: page === 1 ? 'Event' : 'Article',
          author: `Author ${itemNumber % 5 + 1}`,
          status: 'Published'
        });
      }
    }

    console.log(`\nStep 4: Aggregate and trim results`);
    console.log(`- Fetched ${totalFetched} items from ${pagesNeeded} pages`);
    console.log(`- Trimming to ${itemsNeeded} items as requested`);

    const finalContent = allContent.slice(0, itemsNeeded);

    console.log('\n=== Results ===');
    console.log(`Requested: ${itemsNeeded} items`);
    console.log(`Fetched: ${totalFetched} items from ${pagesNeeded} pages`);
    console.log(`Returned: ${finalContent.length} items (trimmed to request)`);

    // Show sample of results
    console.log('\n=== Sample Content ===');
    finalContent.slice(0, 5).forEach((item, index) => {
      console.log(`${index + 1}. ${item.title} (${item.type}) - ${item.author}`);
    });

    if (finalContent.length > 5) {
      console.log(`... and ${finalContent.length - 5} more items`);
    }

    return {
      content: finalContent,
      totalFetched,
      requested: itemsNeeded,
      actual: finalContent.length,
      pagesFetched: pagesNeeded
    };
  }

  /**
   * Example: Fetch all items of a specific type
   */
  async exampleFetchAllOfType(contentType) {
    console.log(`=== Example: Fetch All ${contentType} Items ===`);

    try {
      // First get pagination info for this type
      const pagination = await this.getPaginationInfo(contentType);
      console.log(`Found ${pagination.totalPages} pages of ${contentType} content`);

      const result = await this.fetchItems(
        pagination.totalPages * 50, // Estimate based on pages
        { type: contentType }
      );

      console.log(`Fetched ${result.actual} ${contentType} items total`);
      return result;

    } catch (error) {
      console.error(`Error fetching ${contentType} items:`, error.message);
      throw error;
    }
  }
}

// CLI interface
async function main() {
  const processor = new DrupalContentBatchProcessor();

  try {
    // Example 1: Fetch 75 items (crosses page boundary) - Demonstration
    await processor.exampleFetch75ItemsDemonstration();

    console.log('\n' + '='.repeat(50) + '\n');

    // Example 2: Show how it would work with real API
    console.log('=== Real Implementation Notes ===');
    console.log('To run with actual API calls:');
    console.log('1. Ensure Drupal session is loaded: POST /login/load');
    console.log('2. Run: await processor.exampleFetch75Items()');
    console.log('3. Handle potential browser session rate limiting');
    console.log('4. Consider adding delays between requests');

  } catch (error) {
    console.error('Batch processing demonstration failed:', error.message);
    process.exit(1);
  }
}

// Export for use as module
module.exports = DrupalContentBatchProcessor;

// Run if called directly
if (require.main === module) {
  main();
}